File: 110_balanced_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Time Complexity: O(n ^ 2)
// Space Complexity: O(n)
/*
class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        return Math.abs(getDepth(root.left) - getDepth(root.right)) <= 1
            && isBalanced(root.left)
            && isBalanced(root.right);
    }
    
    private int getDepth(TreeNode node) {
        if (node == null) return 0;
        return Math.max(getDepth(node.left), getDepth(node.right)) + 1;
    }
}
*/

// propagation from bottom to the top 
class Solution {
    public boolean isBalanced(TreeNode root) {
        return getMaxDepth(root) != -1;    
    }
    
    private int getMaxDepth(TreeNode root) {
        if (root == null) return 0;
        
        int left = getMaxDepth(root.left);
        if (left == -1) return -1;
        
        int right = getMaxDepth(root.right);
        if (right == -1) return -1;
        
        return (Math.abs(left - right) <= 1) ? Math.max(left, right) + 1 : -1;
    }
}

File: python/110_balanced_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isBalanced(self, root: TreeNode) -> bool:
        return self.getDepth(root) != -1

    def getDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        left_dep = self.getDepth(root.left)
        if left_dep == -1:
            return -1

        right_dep = self.getDepth(root.right)
        if right_dep == -1:
            return -1

        max_depth = max(left_dep, right_dep)
        return max_depth + 1 if abs(left_dep - right_dep) <= 1 else -1
